### PR TITLE
qemu: Add leon2 support

### PR DESCRIPTION
**Pull request by yashi**
Hi,

This is a patch to add LEON2 to SPARC architecture.   This is need to run SPARC arch PR for Zephyr.  This is one commit having 3 patches.  Each patch is in Git format style and has its own commit log.

This PR is more of a request for comment because I've never added patch to sdk-ng.   Feedback comment will be appreciated.

### CODE REVIEW

**Comment by galak** [comment]
Looks like you're getting a build error:
```
ERROR: ParseError at /home/buildslave/src/github.com/zephyrproject-rtos/sdk-ng/meta-zephyr-sdk/recipes-devtools/qemu/zephyr-qemu_git.bb:14: unparsed line: 'SRC_URI = "git://github.com/qemu/qemu.git;protocol=https 	   file://0001-qemu-nios2-Add-Altera-MAX-10-board-support-for-Zephy.patch 	   file://0001-hw-sparc-Add-leon-at697-machine.patch 	   file://0002-hw-sparc-leon-Fix-compilation-errors.patch 	   file://0003-hw-sparc-leon-timer-Call-leon_timer_io_read-for-TIME.patch'
Loading cache...done.
Loaded 1 entries from dependency cache.
Parsing recipes...ERROR: ParseError at /home/buildslave/src/github.com/zephyrproject-rtos/sdk-ng/meta-zephyr-sdk/recipes-devtools/qemu/zephyr-qemu_git.bb:14: unparsed line: 'SRC_URI = "git://github.com/qemu/qemu.git;protocol=https 	   file://0001-qemu-nios2-Add-Altera-MAX-10-board-support-for-Zephy.patch 	   file://0001-hw-sparc-Add-leon-at697-machine.patch 	   file://0002-hw-sparc-leon-Fix-compilation-errors.patch 	   file://0003-hw-sparc-leon-timer-Call-leon_timer_io_read-for-TIME.patch'

```

File: meta-zephyr-sdk/recipes-devtools/qemu/zephyr-qemu_git.bb [hunk 1]
```diff
@@ -9,6 +9,9 @@ LIC_FILES_CHKSUM = "file://COPYING;md5=441c28d2cf86e15a37fa47e15a72fbac \
 SRCREV = "aa464db69b40b4b695be31085e6d2f1e90956c89"
 SRC_URI = "git://github.com/qemu/qemu.git;protocol=https \
 	   file://0001-qemu-nios2-Add-Altera-MAX-10-board-support-for-Zephy.patch \
+	   file://0001-hw-sparc-Add-leon-at697-machine.patch \
+	   file://0002-hw-sparc-leon-Fix-compilation-errors.patch \
+	   file://0003-hw-sparc-leon-timer-Call-leon_timer_io_read-for-TIME.patch
 "
```
Comment (stephanosio) [inline meta-zephyr-sdk/recipes-devtools/qemu/zephyr-qemu_git.bb]: `file://0003-hw-sparc-leon-timer-Call-leon_timer_io_read-for-TIME.patch \`

File: meta-zephyr-sdk/recipes-devtools/qemu/files/0001-hw-sparc-Add-leon-at697-machine.patch [hunk 2]
```diff
@@ -0,0 +1,978 @@
...
+index 355b07ae05..2856251f66 100644
+--- a/hw/sparc/trace-events
++++ b/hw/sparc/trace-events
+@@ -19,3 +19,16 @@ sun4m_iommu_bad_addr(uint64_t addr) "bad addr 0x%"PRIx64
+ # leon3.c
+ leon3_set_irq(int intno) "Set CPU IRQ %d"
+ leon3_reset_irq(int intno) "Reset CPU IRQ %d"
++
++# hw/leon.c
++leon_check_irqs(uint32_t pend, uint32_t force, uint32_t mask, uint32_t lvl1, uint32_t lvl2) "pend:0x%04x force:0x%04x mask:0x%04x lvl1:0x%04x lvl0:0x%04x\n"
```
Comment (stephanosio) [inline meta-zephyr-sdk/recipes-devtools/qemu/files/0001-hw-sparc-Add-leon-at697-machine.patch]: Maybe `\n` here is the issue?

Comment (yashi) [reply meta-zephyr-sdk/recipes-devtools/qemu/files/0001-hw-sparc-Add-leon-at697-machine.patch]: You mean, the very end of `\n` at `0x%04x\n"`?

Still trying to build it locally but haven't been able to successfully build it yet.  And it's taking ages...

Comment (stephanosio) [reply meta-zephyr-sdk/recipes-devtools/qemu/files/0001-hw-sparc-Add-leon-at697-machine.patch]: > You mean, the very end of `\n` at `0x%04x\n"`?

Yes

**Comment by yashi** [comment]
Thanks @stephanosio.  I think somehow I missed transaction-based ptimer API[1].  I'll fix it in a few days.

[1]: https://github.com/qemu/qemu/commit/91b37aea0e3f344aefec4d72faf12cbcac79d64e

File: meta-zephyr-sdk/recipes-devtools/qemu/files/0001-hw-sparc-Add-leon-at697-machine.patch (hunk 2 above)
Comment (yashi) [reply meta-zephyr-sdk/recipes-devtools/qemu/files/0001-hw-sparc-Add-leon-at697-machine.patch]: Removed.

**Comment by yashi** [comment]
I've addressed all issues and test against QEMU v4.2.0-rc4.  Hope it works.

**Comment by stephanosio** [comment]
@yashi So with these changes, QEMU should be able to run tests for the SPARC arch (in zephyrproject-rtos/zephyr#17978), right?

**Comment by yashi** [comment]
Yes.  At least simple tests under `tests/kernel`.

**Comment by yashi** [comment]
Sure.  I'll push it later.
Just for reference:
0001- is taken from AdaCore as is.
0002- is mine to fix the build failure because 0001 depends on some other commits in the AdaCore repo.
0003- is a bug we found.  I should have reported to AdaCore but haven't.
0004- is a fix needed to rebase on top of v4.2-rc

And yest, I'm planning to talk to AdaCore about leon2 support in QEMU upstream ML.

**Comment by yashi** [comment]
Done updating the commit log.

**Pull request merged**